File: src/indra_belief/curation.py
"""INDRA curations as gold — the single source of truth for the curation domain.

Until now this logic lived, copied, across four scripts (pull_rasmachine_curations,
curation_accuracy, build_disagreement_queue, prepare_dataset) and was re-expressed
again in the SvelteKit viewer's TypeScript. The `tag == "correct"` gold atom was
independently rewritten in a half-dozen eval scripts besides. This module is the
canonical Python home; the viewer's `viewer/src/lib/data/curation.ts` is its
deliberate cross-language twin, kept in parity by tests/test_curation_parity (a
fixture both sides must reduce identically).

Five concepts, one home:

1. The curation record — what INDRA's curation DB returns per (statement, evidence).
2. The GOLD RULE — a curation tag is gold-correct iff it is exactly "correct";
   every other tag (no_relation, wrong_relation, grounding, polarity, act_vs_amt,
   hypothesis, negative_result, entity_boundaries, mod_site, other) means the
   reader's extraction is wrong. When an evidence has several curations,
   AGGREGATE with any-incorrect-wins (one curator objection flips it).
3. The JOIN KEY — curations key on the INDRA-native int pair
   (matches_hash, source_hash). Our run exports store the statement hash as a
   STRING (`indra_matches_hash`); this module owns the str->int coercion so no
   call site re-implements it.
4. The INDEX — a JSONL of curations reduced to {key -> GoldVerdict}.
5. TRANSPORT — fetching curations from the public INDRA DB REST endpoint.

Transport (httpx/async) is imported lazily so the pure domain (1-4) has no
third-party dependency and is trivially testable.
"""
from __future__ import annotations

import json
import os
from collections import defaultdict
from dataclasses import dataclass, field
from typing import Iterable

# ── 1. the curation record ──────────────────────────────────────────────────

#: Every curation tag the INDRA DB emits. Exactly one — "correct" — means the
#: reader's extraction is supported; the rest each name a way it is wrong.
CURATION_TAGS: tuple[str, ...] = (
    "correct",
    "no_relation",
    "wrong_relation",
    "grounding",
    "polarity",
    "act_vs_amt",
    "hypothesis",
    "negative_result",
    "entity_boundaries",
    "agent_conditions",
    "mod_site",
    "other",
)

#: The one tag that denotes a correct extraction. Everything else is "incorrect".
CORRECT_TAG = "correct"


@dataclass(frozen=True)
class Curation:
    """One human curation of a (statement, evidence) extraction.

    matches_hash / source_hash are the INDRA-native int identifiers. The DB
    payload calls the statement key `pa_hash`; our pulled file mirrors it as
    `_matches_hash`. Either is accepted by `from_dict`.
    """

    matches_hash: int
    source_hash: int
    tag: str
    curator: str = ""
    date: str = ""
    text: str = ""

    @classmethod
    def from_dict(cls, d: dict) -> "Curation | None":
        mh = d.get("_matches_hash", d.get("pa_hash"))
        sh = d.get("source_hash")
        if mh is None or sh is None:
            return None
        try:
            mh_i, sh_i = int(mh), int(sh)
        except (TypeError, ValueError):
            return None
        return cls(
            matches_hash=mh_i,
            source_hash=sh_i,
            tag=str(d.get("tag", "")),
            curator=str(d.get("curator", "") or ""),
            date=str(d.get("date", "") or ""),
            text=str(d.get("text", "") or ""),
        )


# ── 2. the gold rule ────────────────────────────────────────────────────────


def is_gold_correct(tag: str | None) -> bool:
    """The gold atom: a single curation tag is correct iff it is exactly "correct".

    This is the one predicate the benchmark-eval scripts each used to rewrite as
    `record["tag"] == "correct"`. Import it instead.
    """
    return tag == CORRECT_TAG


def aggregate_gold(tags: Iterable[str]) -> str | None:
    """Aggregate one evidence's curation tags into a gold verdict.

    any-incorrect-wins: "correct" iff EVERY tag is "correct"; "incorrect" if any
    tag dissents. Returns None for an empty set (uncurated). Conservative by
    design — a lone curator objection is enough to mark the extraction wrong.
    """
    tags = list(tags)
    if not tags:
        return None
    return CORRECT_TAG if all(is_gold_correct(t) for t in tags) else "incorrect"


# ── 3. the join key ─────────────────────────────────────────────────────────


def curation_key(matches_hash, source_hash) -> tuple[int, int] | None:
    """The content-addressed join key, coercing run-export string hashes to int.

    Run exports store the statement key as a string (`indra_matches_hash`) and
    the evidence key as an int (`source_hash`); curations carry both as int.
    Returns None when either hash is missing or unparseable.
    """
    if matches_hash is None or source_hash is None:
        return None
    try:
        return (int(matches_hash), int(source_hash))
    except (TypeError, ValueError):
        return None


# ── 4. the gold index ───────────────────────────────────────────────────────


@dataclass
class GoldVerdict:
    """The aggregated human verdict for one (matches_hash, source_hash)."""

    verdict: str  # "correct" | "incorrect"
    n: int
    tags: list[str]
    curators: list[str] = field(default_factory=list)
    notes: list[str] = field(default_factory=list)


@dataclass
class CurationIndex:
    """Curations indexed by the int join key, with derived gold verdicts.

    The viewer's CurationIndex (store.ts) mirrors this shape. `present` lets a
    caller distinguish "no curations file" from "file present but this evidence
    uncurated" without raising.
    """

    by_key: dict[tuple[int, int], list[Curation]]
    gold_by_key: dict[tuple[int, int], GoldVerdict]
    present: bool = True

    @property
    def n_statements(self) -> int:
        return len({mh for mh, _ in self.by_key})

    @property
    def n_evidences(self) -> int:
        return len(self.by_key)

    def gold_for(self, matches_hash, source_hash) -> GoldVerdict | None:
        """Look up the gold verdict for an evidence by its (possibly string)
        hashes — the canonical replacement for every ad-hoc coerce-and-lookup."""
        key = curation_key(matches_hash, source_hash)
        if key is None:
            return None
        return self.gold_by_key.get(key)

    def is_curated(self, matches_hash, source_hash) -> bool:
        return self.gold_for(matches_hash, source_hash) is not None


EMPTY_INDEX = CurationIndex(by_key={}, gold_by_key={}, present=False)


def build_index(curations: Iterable[Curation]) -> CurationIndex:
    """Reduce curations into a CurationIndex (raw groups + gold verdicts)."""
    by_key: dict[tuple[int, int], list[Curation]] = defaultdict(list)
    for c in curations:
        by_key[(c.matches_hash, c.source_hash)].append(c)

    gold_by_key: dict[tuple[int, int], GoldVerdict] = {}
    for key, curs in by_key.items():
        tags = [c.tag for c in curs]
        verdict = aggregate_gold(tags) or "incorrect"
        curators, seen = [], set()
        for c in curs:
            if c.curator and c.curator not in seen:
                seen.add(c.curator)
                curators.append(c.curator)
        notes = [c.text for c in curs if c.text and c.text.strip()]
        gold_by_key[key] = GoldVerdict(verdict=verdict, n=len(curs), tags=tags, curators=curators, notes=notes)

    return CurationIndex(by_key=dict(by_key), gold_by_key=gold_by_key, present=True)


def load_index(path: str) -> CurationIndex:
    """Load a curations JSONL into a CurationIndex. Missing file -> EMPTY_INDEX
    (present=False), so gold-aware callers degrade honestly rather than crash."""
    if not path or not os.path.exists(path):
        return EMPTY_INDEX
    curations: list[Curation] = []
    with open(path) as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                d = json.loads(line)
            except json.JSONDecodeError:
                continue
            c = Curation.from_dict(d)
            if c is not None:
                curations.append(c)
    return build_index(curations)


# ── 5. transport (lazy: no httpx import unless you pull) ─────────────────────

INDRA_DB_REST_URL = "https://db.indra.bio"
CURATION_LIST_PATH = "/curation/list"


def curation_list_url(matches_hash: int, base_url: str = INDRA_DB_REST_URL) -> str:
    """The public, key-required curation endpoint (no auth needed for by-hash)."""
    return f"{base_url.rstrip('/')}{CURATION_LIST_PATH}/{matches_hash}"


async def fetch_curations(
    matches_hashes: Iterable[int],
    *,
    base_url: str = INDRA_DB_REST_URL,
    concurrency: int = 16,
    retries: int = 4,
    backoff: float = 0.5,
    timeout: float = 30.0,
    on_progress=None,
) -> tuple[list[dict], list[int]]:
    """Async-pull curations for the given statement hashes from the INDRA DB.

    Returns (curation_dicts, failed_hashes). Each returned dict is the raw DB
    payload augmented with `_matches_hash`. Bounded concurrency + connection
    pooling (resolves the host once) + retries-with-backoff — the by-hash route
    is public, no API key. httpx is imported here so the pure domain above has
    no third-party dependency.
    """
    import asyncio

    import httpx

    hashes = list(matches_hashes)
    sem = asyncio.Semaphore(concurrency)
    out: list[dict] = []
    failed: list[int] = []
    limits = httpx.Limits(max_connections=concurrency, max_keepalive_connections=concurrency)

    async def one(client: httpx.AsyncClient, mh: int) -> None:
        async with sem:
            for attempt in range(retries + 1):
                try:
                    r = await client.get(f"{CURATION_LIST_PATH}/{mh}")
                    if r.status_code == 200:
                        for c in r.json():
                            c["_matches_hash"] = mh
                            out.append(c)
                        if on_progress:
                            on_progress(mh, True)
                        return
                    if 400 <= r.status_code < 500:
                        if on_progress:
                            on_progress(mh, True)
                        return
                except (httpx.TransportError, httpx.HTTPError):
                    pass
                if attempt < retries:
                    await asyncio.sleep(backoff * (2**attempt))
            failed.append(mh)
            if on_progress:
                on_progress(mh, False)

    async with httpx.AsyncClient(
        base_url=base_url,
        limits=limits,
        timeout=httpx.Timeout(timeout, connect=timeout),
        headers={"User-Agent": "indra-belief-curation/1"},
    ) as client:
        await asyncio.gather(*(one(client, mh) for mh in hashes))

    return out, failed

File: viewer/src/lib/data/curation.ts
/**
 * INDRA curations as gold — the viewer's canonical curation-domain module, and
 * the deliberate cross-language twin of `src/indra_belief/curation.py`. The gold
 * rule, the (matches_hash, source_hash) join + str->int coercion, the index
 * reduction, and the single evidence->gold lookup all live here ONCE. store.ts
 * owns caching/IO; queries.ts and adjudicate.ts consume `goldForRow`. Parity
 * with the Python module is enforced by tests/test_curation_parity.
 *
 * The gold rule: a curation tag is gold-correct iff it is exactly "correct";
 * every other tag means the reader's extraction is wrong. Multiple curations on
 * one evidence aggregate with any-incorrect-wins (one objection flips it).
 */
import type { CurationRow, GoldVerdict, EvidenceRow } from './types';

/** The one tag denoting a correct extraction. Everything else is "incorrect". */
export const CORRECT_TAG = 'correct';

/** The gold atom: a single curation tag is correct iff it is exactly "correct". */
export function isGoldCorrect(tag: string | null | undefined): boolean {
	return tag === CORRECT_TAG;
}

/** Aggregate one evidence's curation tags into a gold verdict, any-incorrect-wins.
 *  Returns null for an empty set (uncurated). */
export function aggregateGold(tags: string[]): 'correct' | 'incorrect' | null {
	if (tags.length === 0) return null;
	return tags.every(isGoldCorrect) ? 'correct' : 'incorrect';
}

/** Content-addressed join key. Curation hashes are ints; run-export
 *  `indra_matches_hash` is a string — coerce here so no call site repeats it.
 *  Returns null when either hash is missing or unparseable. */
export function curationKey(
	matchesHash: number | string | null | undefined,
	sourceHash: number | string | null | undefined
): string | null {
	if (matchesHash == null || sourceHash == null) return null;
	const mh = typeof matchesHash === 'string' ? parseInt(matchesHash, 10) : Number(matchesHash);
	const sh = typeof sourceHash === 'string' ? parseInt(sourceHash, 10) : Number(sourceHash);
	if (!Number.isFinite(mh) || !Number.isFinite(sh)) return null;
	return `${mh}|${sh}`;
}

/** The in-memory curation index: raw groups + derived gold verdicts, keyed by
 *  `${matches_hash}|${source_hash}`. Mirrors curation.py's CurationIndex. */
export interface CurationIndex {
	byKey: Map<string, CurationRow[]>;
	goldByKey: Map<string, GoldVerdict>;
	nStatements: number;
	nEvidences: number;
	/** true when the curations file was present at all (vs. not pulled). */
	present: boolean;
}

export const EMPTY_CURATION_INDEX: CurationIndex = {
	byKey: new Map(),
	goldByKey: new Map(),
	nStatements: 0,
	nEvidences: 0,
	present: false
};

/** Reduce parsed curation rows into a CurationIndex (raw groups + gold). The
 *  pure core — store.ts wraps it with file IO + mtime caching. */
export function buildCurationIndex(rows: Iterable<CurationRow>): CurationIndex {
	const byKey = new Map<string, CurationRow[]>();
	const statements = new Set<number>();
	for (const c of rows) {
		if (c._matches_hash == null || c.source_hash == null) continue;
		const key = curationKey(c._matches_hash, c.source_hash);
		if (key == null) continue;
		let arr = byKey.get(key);
		if (!arr) byKey.set(key, (arr = []));
		arr.push(c);
		statements.add(Number(c._matches_hash));
	}

	const goldByKey = new Map<string, GoldVerdict>();
	for (const [key, curs] of byKey) {
		const tags = curs.map((c) => c.tag);
		const verdict = aggregateGold(tags) ?? 'incorrect';
		goldByKey.set(key, {
			verdict,
			n: curs.length,
			tags,
			curators: [...new Set(curs.map((c) => c.curator).filter(Boolean))],
			notes: curs.map((c) => c.text).filter((t) => t && t.trim())
		});
	}

	return { byKey, goldByKey, nStatements: statements.size, nEvidences: byKey.size, present: true };
}

/** Look up the gold verdict for an evidence row by coercing its (string)
 *  indra_matches_hash + source_hash to the index key. The single canonical
 *  replacement for the coerce-and-lookup that was written three times across
 *  queries.ts (joinEvidence, evidenceSideBySide) and adjudicate.ts (goldFor). */
export function goldForRow(index: CurationIndex, row: EvidenceRow | null | undefined): GoldVerdict | null {
	if (!row) return null;
	// Per-run baked gold travels with the run and switches when you switch runs.
	// A baked run sets `gold` on every row (object = curated, null = uncurated);
	// only a legacy run leaves it `undefined`, and only then do we consult the
	// global index. This is what makes gold follow the selected run.
	if (row.gold !== undefined) return row.gold;
	const key = curationKey(row.indra_matches_hash, row.source_hash);
	if (key == null) return null;
	return index.goldByKey.get(key) ?? null;
}
